### PR TITLE
Implement usage metrics module and entry point

### PR DESCRIPTION
**Pull request by dmgaldi**
## Overview
* Added a script for ClinEpiUsage metrics
* [Here's the doc container metrics spec:](https://veupathdb.atlassian.net/wiki/spaces/CLIN/pages/119177229/ClinEpiDB+metrics)

Example metrics in dev:

<html xmlns:v="urn:schemas-microsoft-com:vml"
xmlns:o="urn:schemas-microsoft-com:office:office"
xmlns:x="urn:schemas-microsoft-com:office:excel"
xmlns="http://www.w3.org/TR/REC-html40">

<head>

<meta name=ProgId content=Excel.Sheet>
<meta name=Generator content="Microsoft Excel 15">
<link id=Main-File rel=Main-File
href="file:////Users/rooslab/Library/Group%20Containers/UBF8T346G9.Office/TemporaryItems/msohtmlclip/clip.htm">
<link rel=File-List
href="file:////Users/rooslab/Library/Group%20Containers/UBF8T346G9.Office/TemporaryItems/msohtmlclip/clip_filelist.xml">
<style>

</style>
</head>

<body link="#0563C1" vlink="#954F72">

objects_bucket | objects_count | registered_users_with_analysis_count | guest_users_with_analysis_count | guests_users_with_filter_count | registered_users_with_filter_count
-- | -- | -- | -- | -- | --
0 | 0 | 0 | 0 | 974 | 0
1 | 1 | 8 | 974 | 254 | 8
2 | 2 | 2 | 191 | 52 | 2
<=4 | 7 | 8 | 89 | 32 | 8
<=8 | 26 | 2 | 43 | 5 | 2
<=16 | 84 | 1 | 19 | 5 | 1
<=32 | 156 | 1 | 6 | 0 | 1
<=64 | 63 | 1 | 0 | 0 | 1
\>64 | 237 | 2 | 0 | 0 | 2

</body>

</html>

<html xmlns:v="urn:schemas-microsoft-com:vml"
xmlns:o="urn:schemas-microsoft-com:office:office"
xmlns:x="urn:schemas-microsoft-com:office:excel"
xmlns="http://www.w3.org/TR/REC-html40">

<head>

<meta name=ProgId content=Excel.Sheet>
<meta name=Generator content="Microsoft Excel 15">
<link id=Main-File rel=Main-File
href="file:////Users/rooslab/Library/Group%20Containers/UBF8T346G9.Office/TemporaryItems/msohtmlclip/clip.htm">
<link rel=File-List
href="file:////Users/rooslab/Library/Group%20Containers/UBF8T346G9.Office/TemporaryItems/msohtmlclip/clip_filelist.xml">

</head>

<body link="#0563C1" vlink="#954F72">

</body>

</html>

<html xmlns:v="urn:schemas-microsoft-com:vml"
xmlns:o="urn:schemas-microsoft-com:office:office"
xmlns:x="urn:schemas-microsoft-com:office:excel"
xmlns="http://www.w3.org/TR/REC-html40">

<head>

<meta name=ProgId content=Excel.Sheet>
<meta name=Generator content="Microsoft Excel 15">
<link id=Main-File rel=Main-File
href="file:////Users/rooslab/Library/Group%20Containers/UBF8T346G9.Office/TemporaryItems/msohtmlclip/clip.htm">
<link rel=File-List
href="file:////Users/rooslab/Library/Group%20Containers/UBF8T346G9.Office/TemporaryItems/msohtmlclip/clip_filelist.xml">
<style>

</style>
</head>

<body link="#0563C1" vlink="#954F72">

  | AVENIR-1 | JILCOST-1
-- | -- | --
2/17/23 | 0 | 1
2/18/23 | 0 | 0

</body>

</html>

### CODE REVIEW

**Review by Foxcapades — APPROVED** [review]

File: bin/run.py [hunk 1]
```diff
@@ -0,0 +1,25 @@
...
+import sys
+
+import usagemetrics.runner as runner
+
+
+def __main__():
+    auth_key = os.environ.get('AUTH_KEY')
+
+    if len(sys.argv) != 4:
+        print(f"Usage: {sys.argv[0]} <ENV> <EDA_URL> <PROMETHEUS_URL> <CALENDAR_MONTH|YYYY-MM>")
```
Comment (steve-fischer-200) [inline bin/run.py:11]: either here or in readme give an example usage.  (i dont know what <ENV> is)

File: src/usagemetrics/prometheus_metrics_client.py [hunk 2]
```diff
@@ -0,0 +1,39 @@
...
+        prom_client = client.HTTPConnection(self.base_url)
+        prom_client.request(method="GET", url=url)
+        response = prom_client.getresponse()
+        if response.status != 200:
+            raise RuntimeError("Prometheus service did not return a successful response. " + str(response.read()))
+        df = self.parse_to_dataframe(response, labels)
+        return df
+
+    # Takes a response from the prometheus client and returns a dataframe with a column multi index on the provided
+    # labels and a row index of data point times.
```
Comment (steve-fischer-200) [inline src/usagemetrics/prometheus_metrics_client.py:25]: i still don't understand this :(

File: src/usagemetrics/runner.py [hunk 3]
```diff
@@ -0,0 +1,85 @@
...
+class UsageMetricsRunner:
+    PROJECT_ID = "ClinEpiDB"
+
+    def __init__(self, user_metrics_url, auth_key, prometheus_url, env, calendar_month):
+        self.user_metrics_url = user_metrics_url
+        self.auth_key = auth_key
+        self.prometheus_url = prometheus_url
+        self.env = env
+        self.metrics_writer = MetricsWriter(None, None, None)
+        self.prometheus_client = PrometheusClient(self.prometheus_url)
```
Comment (steve-fischer-200) [inline src/usagemetrics/runner.py:21]: seems like PrometheusClient could be in a lib repo?

File: src/usagemetrics/runner.py [hunk 4]
```diff
@@ -0,0 +1,85 @@
...
+    def handle_download_metrics(self):
+        interval = (self.end - self.start).days
+        download_metrics = self.prometheus_client.get_metrics(
+            'increase(dataset_download_requested_total{environment="' + self.env + '"}[' + str(interval) + 'd])',
+            start_date=self.start,
+            end_date=self.end,
+            labels=['user', 'study'])
+
+        # Filter out zeroes before counting users per study
+        download_metrics = download_metrics[download_metrics != 0.0]
```
Comment (steve-fischer-200) [inline src/usagemetrics/runner.py:44]: why not `< 0`

File: src/usagemetrics/user_metrics_client.py [hunk 5]
```diff
@@ -0,0 +1,79 @@
+import json
+import sys
+from http import client
+
+import pandas
+import pandas as pd
+
+
+class UserMetricsClient:
```
Comment (steve-fischer-200) [inline src/usagemetrics/user_metrics_client.py:9]: maybe call this EdaUserServiceMetricsClient?

File: src/usagemetrics/user_metrics_client.py [hunk 6]
```diff
@@ -0,0 +1,79 @@
...
+
+    def __init__(self, url, auth_key, project_id):
+        self.url = url
+        self.auth_key = auth_key
+        self.project_id = project_id
+
+    # Queries the user service and returns a dataframe with a column index on object type and a row index on metric
```
Comment (steve-fischer-200) [inline src/usagemetrics/user_metrics_client.py:16]: there are a couple of other little dataframes we're supposed to output.  it seems maybe they are missing?

**Review by steve-fischer-200 — COMMENTED** [review]
i know this is already merged, but i left some comments.

**Pull request merged**